src/libraries/System.Text.Json/src/System/Text/Json/Serialization/Metadata/PolymorphicTypeResolver.cs

```diff
@@ -231,6 +231,71 @@ public bool TryGetDerivedJsonTypeInfo(object typeDiscriminator, [NotNullWhen(tru
             return result;
         }
 
+        /// <summary>
+        /// Walks the type hierarchy above the current type for any types that use polymorphic configuration.
+        /// </summary>
+        [UnconditionalSuppressMessage("ReflectionAnalysis", "IL2075:UnrecognizedReflectionPattern",
+            Justification = "The call to GetInterfaces will cross-reference results with interface types " +
+                            "already declared as derived types of the polymorphic base type.")]
+        internal static JsonTypeInfo? FindNearestPolymorphicBaseType(JsonTypeInfo typeInfo)
+        {
+            Debug.Assert(typeInfo.IsConfigured);
+
+            if (typeInfo.PolymorphismOptions != null)
+            {
+                // Type defines its own polymorphic configuration.
+                return null;
+            }
+
+            JsonTypeInfo? matchingResult = null;
+
+            // First, walk up the class hierarchy for any supported types.
+            for (Type? candidate = typeInfo.Type.BaseType; candidate != null; candidate = candidate.BaseType)
+            {
+                JsonTypeInfo? candidateInfo = typeInfo.Options.GetTypeInfoInternal(candidate, ensureNotNull: null);
+                if (candidateInfo?.PolymorphismOptions != null)
+                {
+                    // stop on the first ancestor that has a match
+                    matchingResult = candidateInfo;
+                    break;
+                }
+            }
+
+            // Now, walk the interface hierarchy for any polymorphic interface declarations.
+            foreach (Type interfaceType in typeInfo.Type.GetInterfaces())
```

Comment:
If I'm reading the comment from the calling property for this method (AncestorPolymorphicType) correctly, this is only ever used when serializing (not deserializing) an actual object value (ie. not through a generic overload with some type parameter), and never in any other case? Just want to double check whether I should do additional validation otherwise in the Store, since I'm worried these reflection calls here wouldn't play well (since the linker can just nuke all reflection metadata completely there). If this is only used in that case though (which we never use) then we wouldn't possibly be affected by this 🙂

Reply:
You're right it's only used for `object` value serialization. FWIW `GetInterfaces()` is being used in a few other places already, we generally regard it as linker-safe since we only care about the types that the linker hasn't removed. If reflection itself has been trimmed away however I'd be surprised if this code worked.